# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender_api/src/api.rs
# Commit: a0daec0e7301
# Full Hash: a0daec0e73011a6b972d2bc87c1cf620d569af80
# Author: Kris Taeleman <ktaeleman@mozilla.com>
# Date: 2019-09-27 04:28:18
# Regressor Bug: 1582565
# File Overlap Count: 6
# Description:
#   Bug 1582565 - Log transactions into readable format. r=jrmuizel
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D46657
# ==============================================================================

diff -r 3bd01ec4a9e9 -r a0daec0e7301 gfx/wr/webrender_api/src/api.rs
--- a/gfx/wr/webrender_api/src/api.rs	Wed Sep 25 20:46:24 2019 +0000
+++ b/gfx/wr/webrender_api/src/api.rs	Thu Sep 26 16:43:05 2019 +0000
@@ -39,6 +39,35 @@
     DeleteFontInstance(font::FontInstanceKey),
 }
 
+impl fmt::Debug for ResourceUpdate {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            ResourceUpdate::AddImage(ref i) => f.write_fmt(format_args!(
+                "ResourceUpdate::AddImage size({:?})",
+                &i.descriptor.size
+            )),
+            ResourceUpdate::UpdateImage(ref i) => f.write_fmt(format_args!(
+                "ResourceUpdate::UpdateImage size({:?})",
+                &i.descriptor.size
+            )),
+            ResourceUpdate::AddBlobImage(ref i) => f.write_fmt(format_args!(
+                "ResourceUpdate::AddBlobImage size({:?})",
+                &i.descriptor.size
+            )),
+            ResourceUpdate::UpdateBlobImage(i) => f.write_fmt(format_args!(
+                "ResourceUpdate::UpdateBlobImage size({:?})",
+                &i.descriptor.size
+            )),
+            ResourceUpdate::DeleteImage(..) => f.write_str("ResourceUpdate::DeleteImage"),
+            ResourceUpdate::SetBlobImageVisibleArea(..) => f.write_str("ResourceUpdate::SetBlobImageVisibleArea"),
+            ResourceUpdate::AddFont(..) => f.write_str("ResourceUpdate::AddFont"),
+            ResourceUpdate::DeleteFont(..) => f.write_str("ResourceUpdate::DeleteFont"),
+            ResourceUpdate::AddFontInstance(..) => f.write_str("ResourceUpdate::AddFontInstance"),
+            ResourceUpdate::DeleteFontInstance(..) => f.write_str("ResourceUpdate::DeleteFontInstance"),
+        }
+    }
+}
+
 /// A Transaction is a group of commands to apply atomically to a document.
 ///
 /// This mechanism ensures that:
@@ -468,6 +497,29 @@
     pub notifications: Vec<NotificationRequest>,
 }
 
+impl fmt::Debug for TransactionMsg {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        writeln!(f, "threaded={}, genframe={}, invalidate={}, low_priority={}",
+                        self.use_scene_builder_thread,
+                        self.generate_frame,
+                        self.invalidate_rendered_frame,
+                        self.low_priority,
+                    ).unwrap();
+        for scene_op in &self.scene_ops {
+            writeln!(f, "\t\t{:?}", scene_op).unwrap();
+        }
+
+        for frame_op in &self.frame_ops {
+            writeln!(f, "\t\t{:?}", frame_op).unwrap();
+        }
+
+        for resource_update in &self.resource_updates {
+            writeln!(f, "\t\t{:?}", resource_update).unwrap();
+        }
+        Ok(())
+    }
+}
+
 impl TransactionMsg {
     pub fn is_empty(&self) -> bool {
         !self.generate_frame &&
@@ -1088,6 +1140,7 @@
         const DISABLE_TEXT_PRIMS = 1 << 22;
         const DISABLE_GRADIENT_PRIMS = 1 << 23;
         const OBSCURE_IMAGES = 1 << 24;
+        const LOG_TRANSACTIONS = 1 << 25;
     }
 }
 